modApiServer/src/org/aion/api/server/nrgprice/NrgBlockPriceStrategy.java

```diff
@@ -0,0 +1,139 @@
+    }
+
+    @Override
+    // in order to have good recommendations, we try to keep the blkPriceQ full
+    public boolean isHungry() {
+        return (blkPriceQ.remainingCapacity() > 0);
+    }
+
+    // notion of "block price" = lowest gas price for all transactions in a block, exluding miner's own transactions
+    // returns null if block is empty, invalid input, block filled only with miner's own transactions
+    private Long getBlkPrice(AionBlock blk) {
+        if (blk == null)
+            return null;
+
+        List<AionTransaction> txns = blk.getTransactionsList();
+        Address coinbase = blk.getCoinbase();
+
+        long[] nrgPrices = new long[txns.size()];
+        int i = 0;
+        for(AionTransaction txn : txns) {
+            if (coinbase.compareTo(txn.getSignature().getPubkey(null)) != 0) {
+                nrgPrices[i++] = txn.getNrgPrice();
+            }
+        }
+
+        // we didn't find any non-miner transactions
+        if (i == 0)
+            return null;
+
+        // Arrays.sort by default sorts in ascending order
+        Arrays.sort(nrgPrices);
+
+        return nrgPrices[0];
+    }
+
+    /* Onus on the holder of an NrgPriceAdvisor instance to provide guarantees on:
+    * 1) Blocks provided in the right order
+    * 2) No duplicate blocks provided
+    *
+    * No mechanism anywhere to invalidate computed blkPrice values here (in case of chain re-orgs, etc.)
+    * Rationale: if the block was mined and accepted by some part of the network that included this node,
+    * then that block represents work done by the network; the person who did the work should have made the
+    * greediest gasPrice decisions wrt. including transactions in the block. Therefore, those side-chained
+    * blocks are valid signals by miners of acceptable nrg prices
+    */
+    @Override
+    public void processBlock(AionBlock blk) {
+        Long blkPrice = getBlkPrice(blk);
+        if (blkPrice != null) {
+            if (!blkPriceQ.offer(blkPrice)) {
+                blkPriceQ.poll();
+                if (!blkPriceQ.offer(blkPrice))
+                    LOG.error("NrgBlockPriceStrategy - problem with backing queue implementation");
+                    // alternatively, we could throw here
+            }
+        }
+    }
+
+    @Override
+    public long computeRecommendation() {
+        // if I'm still hungry, then I can't give a good enough prediction yet.
+        // if I'm still hungry, and if the chain is being supported by proof of work, the miners will accept
+        // transaction with any gasPrice > some minimum threshold they've set internally.
+        if (windowSize == 0 || isHungry())
+            return defaultPrice;
+
+        // let the backing syncronized collection do the locking for us.
+        Long[] blkPrice = blkPriceQ.toArray(new Long[blkPriceQ.size()]);
+        Arrays.sort(blkPrice);
+
+        long recommendation = blkPrice[recommendationIndex];
```

Comment:
As noted by @ali-sharif , we take the percentile to be the index offset, this does not set any guarantees on the value of the index relative to the rest of the lislist, a further enhancement could be to take an average of adjacent indices (or some other operation to obtain a consistent value)

Reply:
adding this as an enhancement ticket.

Reply:
https://github.com/aionnetwork/aion/issues/237